### ISSUE TITLE
Better integration with {epiparameter}

### ISSUE THREAD

**Issue by joshwlambert** [epiverse-trace/superspreading#32]
Currently {superspreading} uses data from {epiparameter} in the get started vignette. However, it would be easier if the `<epidist>` object could be passed directly to the functions in {superspreading} that require the parameters of the offspring distribution. Functions that require $R$ and $k$ could have an `epidist` argument which the user could supply instead of the parameters.

This feature would enable better interoperability between the two Epiverse-TRACE packages.

**Comment by joshwlambert** [comment]
I do not want to remove the `R` and `k` parameter arguments from functions as these are easy to supply for everyone using the package. In the case that an `<epidist>` object is supplied it should warn/error if `R` and `k` are supplied. 

The precedence in situations where both are supplied is hard to establish, but I will assume that if a user has supplied an `<epidist>` object then this is what they intend to use, so will use the parameters from `<epidist>`. 

Pertinent info on this topic: https://design.tidyverse.org/args-independence.html

**Comment by joshwlambert** [comment]
https://design.tidyverse.org/def-required.html#what-are-the-exceptions-1 provides good examples of where the pattern of argument interactions are used.

**Comment by joshwlambert** [comment]
Two possible function designs: 

1. `probability_epidemic <- function(R, k, num_init_infect, ..., epidist = NULL)`
2. `probability_epidemic <- function(params, num_init_infect)`

Benefits of first method is the parameters are explicit arguments and only if the user specifies an `<epidist>` object is it used. Downsides of first method is there is an interaction between parameters which can make the function harder to use.

Benefits of second method are simpler function interface and no argument interaction. Downsides are that the user now has to provide a named vector with the parameters or an `<epidist>` object to `params`. It is also not clear from the argument name `params` that an `<epidist>` object is permitted and requires users to go to the documentation.

**Comment by jamesmbaazam** [comment]
A third option would be to rather make `epidist` the first argument and rather make `R`and `k` NULL. The goal here is to encourage interoperability with {epiparameter}.

**Comment by joshwlambert** [comment]
@jamesmbaazam Thanks, it's definitely a viable option. The downside is that it forces users to know what an `<epidist>` object is, which for new users will be esoteric. The best balance will be giving newer users functions that are easy to use, while facilitating more experienced users the ability to pass objects (like `<epidist>`) through the pipeline.

**Comment by pratikunterwegs** [comment]
I agree with this effort, and I've mentioned this in my review as well. There's a benefit to allow passing either an `<epidist>` or `R, k`. You'd likely need to include functions that convert between the information contained in `<epidists>` and what each function needs internally.

I'm currently thinking about how to integrate {epiparameter} with {epidemics} ([per this issue](https://github.com/epiverse-trace/epidemics/issues/20)), and having conversion functions from `<epidist>` to say, transmission rate $\beta$ (and/or $R$) that could be passed to the `<infection>` class, could be helpful for that integration also.

**Comment by joshwlambert** [comment]
@pratikunterwegs Thanks! I'm definitely working on making this change. What are your thoughts on the implementation design? I mention some of the tradeoffs above in the issue and am favouring:

`probability_epidemic <- function(params, num_init_infect)`

but this would likely require taking on {epiparameter} as a dependency to handle multiple types of `params` input by the user.

**Comment by pratikunterwegs** [comment]
Thanks, I preferred James' solution of
`p_epidemic = function(epi_dist, num_init_infect, r0 = NULL, k = NULL)`
with checks if both an `<epidist>` and `r0, k` are passed (Should it error or warn? Unsure.). I'm not sure there would be many other kinds of parameterisations, so this might be a good starting point.

I would agree with taking on {epiparameter} as a dependency, although iirc that does bring up the issue of delaying a CRAN release until {epiparameter} is on CRAN.

**Comment by pratikunterwegs** [comment]
One additional point that I didn't make clearly is that if functions to convert `<epidist>`s are being written they should probably live in {epiparameter}.

**Comment by pratikunterwegs** [comment]
> > `function(epi_dist, num_init_infect, r0 = NULL, k = NULL)` with checks if both an  and r0, k are passed
> 
> Okay, but my preference would be to order the arguments the other way around. I think that we lose generality of the functions if we make `<epidist>` the primary object to be passed in and force users to make these -- at present little-known -- objects.
> 
> The style of the functions are currently similar to {finalsize}. My thoughts are that we should be aiming for "hidden interoperability" (i.e. facilitate piping between tools without forcing people to think in an Epiverse way). Let me know if you disagree, but then an argument could be made for making `<epidist>` the first argument of functions like `finalsize::final_size()`.

I think both are good points - your argument order sounds good, and eventually, integrating {epiparameter} and {finalsize} could be an easy first demonstration of new functionality in {epiparameter}.

**Comment by joshwlambert** [comment]
Thanks for all the input. The implementation chosen was 

`probability_epidemic <- function(R, k, num_init_infect, ..., epidist)`

All changes can be seen in PR #52.

> if functions to convert <epidist>s are being written they should probably live in {epiparameter}.

I decided, for now at least, that the utility function required for this change (`get_epidist_param()`) would live in {superspreading}. This can be moved if it becomes clear that other packages interfacing with {epiparameter} need this function as well.

**Comment by joshwlambert** [comment]
Closing this issue after merging #52.

**Comment by joshwlambert** [comment]
Reopening this issue after discussion around Epiverse interoperability (see also epiverse-trace/cfr#59).

@Bisaloo let me know if you think the current implementation of {superspreading} functions accepting `<epidist>` obejcts should be updated.
